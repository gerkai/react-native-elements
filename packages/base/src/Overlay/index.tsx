import { withTheme } from '../config';
import { Overlay, OverlayProps } from './Overlay';

export { Overlay };
export type { OverlayProps };
export default withTheme(Overlay, 'Overlay');
