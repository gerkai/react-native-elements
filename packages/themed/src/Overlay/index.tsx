import { withTheme } from '../config';
import {
  Overlay,
  OverlayProps,
} from '@react-native-elements/base/dist/Overlay/Overlay';

export { Overlay };
export type { OverlayProps };
export default withTheme(Overlay, 'Overlay');
