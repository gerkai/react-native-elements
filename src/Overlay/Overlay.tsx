import React from 'react';
import {
  View,
  StyleSheet,
  Platform,
  TouchableWithoutFeedback,
  Modal,
  ModalProps,
  ViewStyle,
  StyleProp,
} from 'react-native';
import { RneFunctionComponent } from '../helpers';

export type OverlayProps = ModalProps & {
  /** If true, the overlay is visible. */
  isVisible: boolean;

  /** Style of the backdrop container. */
  backdropStyle?: StyleProp<ViewStyle>;

  /** Style of the actual overlay. */
  overlayStyle?: StyleProp<ViewStyle>;

  /** Handler for backdrop press (only works when `fullscreen` is false). */
  onBackdropPress?(): void;

  /** If set to true, the modal will take up the entire screen width and height. */
  fullScreen?: boolean;

  /** Override React Native `Modal` component (usable for web-platform). */
  ModalComponent?: typeof React.Component;
};

/** The Overlay is a view that floats above an appâ€™s content.
 * Overlays are an easy way to inform or request information from the user. */
export const Overlay: RneFunctionComponent<OverlayProps> = ({
  children,
  backdropStyle,
  overlayStyle,
  onBackdropPress = () => null,
  fullScreen = false,
  ModalComponent = Modal,
  isVisible,
  ...rest
}) => (
  <ModalComponent
    visible={isVisible}
    onRequestClose={onBackdropPress}
    transparent
    {...rest}
  >
    <TouchableWithoutFeedback
      onPress={onBackdropPress}
      testID="RNE__Overlay__backdrop"
    >
      <View
        testID="backdrop"
        style={StyleSheet.flatten([styles.backdrop, backdropStyle])}
      />
    </TouchableWithoutFeedback>

    <View style={styles.container} pointerEvents="box-none">
      <View
        style={StyleSheet.flatten([
          styles.overlay,
          fullScreen && styles.fullscreen,
          overlayStyle,
        ])}
      >
        {children}
      </View>
    </View>
  </ModalComponent>
);

const styles = StyleSheet.create({
  backdrop: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, .4)',
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  fullscreen: {
    width: '100%',
    height: '100%',
  },
  overlay: {
    backgroundColor: 'white',
    borderRadius: 3,
    padding: 10,
    ...Platform.select({
      android: {
        elevation: 2,
      },
      default: {
        shadowColor: 'rgba(0, 0, 0, .3)',
        shadowOffset: { width: 0, height: 1 },
        shadowRadius: 4,
      },
    }),
  },
});

Overlay.displayName = 'Overlay';
