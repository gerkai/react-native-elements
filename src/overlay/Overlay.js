import React from 'react';
import PropTypes from 'prop-types';
import {
  View,
  StyleSheet,
  Dimensions,
  Platform,
  TouchableWithoutFeedback,
} from 'react-native';

import { ViewPropTypes, withTheme } from '../config';

const dimensions = Dimensions.get('window');
const windowWidth = dimensions.width;
const windowHeight = dimensions.height;

const Overlay = props => {
  const {
    children,
    isVisible,
    containerStyle,
    overlayStyle,
    windowBackgroundColor,
    overlayBackgroundColor,
    onBackdropPress,
    borderRadius,
    width,
    height,
    fullScreen,
    ...rest
  } = props;

  if (!isVisible) return null;

  return (
    <TouchableWithoutFeedback onPress={onBackdropPress}>
      <View
        testID="overlayContainer"
        style={StyleSheet.flatten([
          styles.container,
          { backgroundColor: windowBackgroundColor },
          containerStyle,
        ])}
        {...rest}
      >
        <TouchableWithoutFeedback>
          <View
            style={StyleSheet.flatten([
              styles.overlay,
              {
                borderRadius,
                backgroundColor: overlayBackgroundColor,
                width,
                height,
              },
              fullScreen && { width: windowWidth, height: windowHeight },
              overlayStyle,
            ])}
          >
            {children}
          </View>
        </TouchableWithoutFeedback>
      </View>
    </TouchableWithoutFeedback>
  );
};

Overlay.propTypes = {
  children: PropTypes.any,
  isVisible: PropTypes.bool,
  containerStyle: ViewPropTypes.style,
  overlayStyle: ViewPropTypes.style,
  windowBackgroundColor: PropTypes.string,
  overlayBackgroundColor: PropTypes.string,
  onBackdropPress: PropTypes.func,
  borderRadius: PropTypes.number,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  fullScreen: PropTypes.bool,
};

Overlay.defaultProps = {
  borderRadius: 3,
  fullScreen: false,
  windowBackgroundColor: 'rgba(0, 0, 0, .4)',
  overlayBackgroundColor: 'white',
  width: windowWidth - 80,
  height: windowHeight - 180,
  onBackdropPress: () => null,
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: windowWidth,
    height: windowHeight,
    justifyContent: 'center',
    alignItems: 'center',
    elevation: 5,
  },
  overlay: {
    borderRadius: 5,
    padding: 10,
    ...Platform.select({
      android: {
        elevation: 2,
      },
      default: {
        shadowColor: 'rgba(0, 0, 0, .3)',
        shadowOffset: { width: 0, height: 1 },
        shadowRadius: 4,
      },
    }),
  },
});

export { Overlay };
export default withTheme(Overlay, 'Overlay');
