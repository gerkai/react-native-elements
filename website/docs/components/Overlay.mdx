---
id: overlay

title: Overlay
---

import Usage from "./usage/Overlay.mdx";

import { Overlay } from "react-native-elements";

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

The Overlay is a view that floats above an appâ€™s content.

Overlays are an easy way to inform or request information from the user.

## Usage

<Usage />

## Props

### Overlay

- [ModalComponent](#modalcomponent)
- [backdropStyle](#backdropstyle)
- [fullScreen](#fullscreen)
- [isVisible](#isvisible)
- [onBackdropPress](#onbackdroppress)
- [onLongPress](#onlongpress)
- [onPressIn](#onpressin)
- [onPressOut](#onpressout)
- [overlayStyle](#overlaystyle)
- [pressableProps](#pressableprops)

## Reference

### Overlay

#### ModalComponent

Override React Native `Modal` component (usable for web-platform).

| Type | Default |
| ---- | ------- |
| any  | None    |

---

#### backdropStyle

Style of the backdrop container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### fullScreen

If set to true, the modal will take up the entire screen width and height.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### isVisible

If true, the overlay is visible.

| Type    | Default |
| ------- | ------- |
| boolean | None    |

---

#### onBackdropPress

Handler for backdrop press (only works when `fullscreen` is false).

| Type     | Default  |
| -------- | -------- |
| Function | Function |

---

#### onLongPress

Called when a long-tap gesture is detected.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPressIn

Called when a touch is engaged before `onPress`.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPressOut

Called when a touch is released before `onPress`.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### overlayStyle

Style of the actual overlay.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### pressableProps

| Type                                 | Default |
| ------------------------------------ | ------- |
| PressableProps except click handlers | None    |

---
