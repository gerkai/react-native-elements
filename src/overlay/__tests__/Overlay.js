import React from 'react';
import { Text } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Overlay from '../Overlay';

describe('Overlay', () => {
  it('should render without issues', () => {
    const component = shallow(
      <Overlay isVisible>
        <Text>I'm in an Overlay</Text>
      </Overlay>
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render nothing if not visible', () => {
    const component = shallow(
      <Overlay isVisible={false}>
        <Text>I'm in an Overlay</Text>
      </Overlay>
    );

    expect(component.getElement()).toBeFalsy();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should be able to render fullscreen', () => {
    const component = shallow(
      <Overlay isVisible={true} fullScreen>
        <Text>I'm in an Overlay</Text>
      </Overlay>
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
