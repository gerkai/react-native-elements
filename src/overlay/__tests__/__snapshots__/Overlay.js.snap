// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should apply values from theme 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "green",
      "elevation": 5,
      "height": 1334,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": 750,
    }
  }
  theme={
    Object {
      "Overlay": Object {
        "windowBackgroundColor": "green",
      },
      "colors": Object {
        "disabled": "hsl(208, 8%, 90%)",
        "divider": "#bcbbc1",
        "error": "#ff190c",
        "grey0": "#393e42",
        "grey1": "#43484d",
        "grey2": "#5e6977",
        "grey3": "#86939e",
        "grey4": "#bdc6cf",
        "grey5": "#e1e8ee",
        "greyOutline": "#bbb",
        "platform": Object {
          "android": Object {
            "error": "#f44336",
            "primary": "#2196f3",
            "secondary": "#9C27B0",
            "success": "#4caf50",
            "warning": "#ffeb3b",
          },
          "ios": Object {
            "error": "#ff3b30",
            "primary": "#007aff",
            "secondary": "#5856d6",
            "success": "#4cd964",
            "warning": "#ffcc00",
          },
        },
        "primary": "#2089dc",
        "searchBg": "#303337",
        "secondary": "#8F0CE8",
        "success": "#52c41a",
        "warning": "#faad14",
      },
    }
  }
  updateTheme={[Function]}
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 3,
        "height": 1154,
        "padding": 10,
        "shadowColor": "rgba(0, 0, 0, .3)",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowRadius": 4,
        "width": 670,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      I'm in an Overlay
    </Text>
  </View>
</View>
`;

exports[`Overlay should be able to render fullscreen 1`] = `
<TouchableWithoutFeedback
  onPress={[Function]}
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, .4)",
        "elevation": 5,
        "height": 1334,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 750,
      }
    }
    testID="overlayContainer"
  >
    <TouchableWithoutFeedback>
      <Component
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 3,
            "height": 1334,
            "padding": 10,
            "shadowColor": "rgba(0, 0, 0, .3)",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowRadius": 4,
            "width": 750,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          I'm in an Overlay
        </Text>
      </Component>
    </TouchableWithoutFeedback>
  </Component>
</TouchableWithoutFeedback>
`;

exports[`Overlay should render nothing if not visible 1`] = `""`;

exports[`Overlay should render without issues 1`] = `
<TouchableWithoutFeedback
  onPress={[Function]}
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, .4)",
        "elevation": 5,
        "height": 1334,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 750,
      }
    }
    testID="overlayContainer"
  >
    <TouchableWithoutFeedback>
      <Component
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 3,
            "height": 1154,
            "padding": 10,
            "shadowColor": "rgba(0, 0, 0, .3)",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowRadius": 4,
            "width": 670,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          I'm in an Overlay
        </Text>
      </Component>
    </TouchableWithoutFeedback>
  </Component>
</TouchableWithoutFeedback>
`;
